File: two_pointers/02_two_sum.py

def two_sum(numbers: list[int], target: int) -> list[int]:
    # assuming input is sorted in non-decreasing order
    i = 0
    j = len(numbers) - 1
    while i + 1 < j:
        while numbers[i] + numbers[j] > target:
            j -= 1
        while numbers[i] + numbers[j] < target:
            i += 1
        if numbers[i] + numbers[j] == target:
            return [i, j]
    return [i, j]
        

def test_two_sum():
    numbers = [1, 2, 3, 4]
    target = 3
    expected = [0, 1]
    assert(two_sum(numbers, target) == expected)

    numbers = [2, 3, 4]
    target = 6
    expected = [0, 2]
    assert(two_sum(numbers, target) == expected)

    numbers = [3, 3]
    target = 6
    expected = [0, 1]
    assert(two_sum(numbers, target) == expected)


if __name__ == "__main__":
    test_two_sum()

File: 02_two_pointers/02_two_sum.cpp
#include <cassert>
#include <vector>
#include <iostream>

using std::vector;

vector<int> two_sum(vector<int>& numbers, int target)
{
	int i = 0;
	int j = numbers.size() - 1;

	while (i < j) {
		int sum = numbers[i] + numbers[j];

		if (sum > target) j--;
		else if (sum < target) i++;
		else return vector<int>{i++, j++};
	}

	return vector<int>{i, j};
}

void test_two_sum()
{
	vector<int> numbers, expected;
	int target;

	numbers = {1, 2, 3, 4};
	target = 3;
	expected = {0, 1};
	assert(two_sum(numbers, target) == expected);

	numbers = {2, 3, 4};
	target = 6;
	expected = {0, 2};
	assert(two_sum(numbers, target) == expected);

	numbers = {3, 3};
	target = 6;
	expected = {0, 1};
	assert(two_sum(numbers, target) == expected);

	std::cout << "All tests passed" << std::endl;
}

int main()
{
	test_two_sum();
	return 0;
}
